tests/fuzz/s2n_server_hello_recv_test.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* Target Functions: s2n_server_hello_recv s2n_server_extensions_recv s2n_server_hello_retry_recv
+                     s2n_connection_get_cipher_preferences s2n_set_cipher_as_client */
+
+#include <stdint.h>
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_tls.h"
+#include "utils/s2n_safety.h"
+#include "s2n_test.h"
+
+static const uint8_t TLS_VERSIONS[] = {S2N_TLS10, S2N_TLS11, S2N_TLS12, S2N_TLS13};
+struct s2n_config *client_config;
+
+static void s2n_fuzz_atexit()
+{
+    s2n_cleanup();
+}
+
+int LLVMFuzzerInitialize(const uint8_t *buf, size_t len)
+{
+#ifdef S2N_TEST_IN_FIPS_MODE
+    S2N_TEST_ENTER_FIPS_MODE();
+#endif
+
+    GUARD(s2n_init());
+    GUARD_STRICT(atexit(s2n_fuzz_atexit));
+    return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    /* We need at least one byte of input to set parameters */
+    if (len != 0) {
+        /* Setup */
+        client_config = s2n_config_new();
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        GUARD(s2n_connection_set_config(client_conn, client_config));
+        notnull_check(client_conn);
+        GUARD(s2n_stuffer_write_bytes(&client_conn->handshake.io, buf, len));
+
+        /* Pull a byte off the libfuzzer input and use it to set parameters */
+        uint8_t randval = 0;
+        GUARD(s2n_stuffer_read_uint8(&client_conn->handshake.io, &randval));
```

Comment:
Same comment as before, I'd prefer a for loop instead of letting LibFizzer guess a value. This will work, but it means you will be storing many more corpus files and LibFuzzer might miss some interesting inputs.

Reply:
After speaking with Audrey offline, it seems that doing it this way actually increases LibFuzzer code coverage. Our current theory is that performing multiple test runs in a for loop in a single fuzz test leads to noisier coverage measurements for LibFuzzer, which makes it harder for LibFuzzer figure out what byte change resulted in which branch coverage change.